backstory/src/main/java/com/google/sps/servlets/AnalyzeImageServlet.java

```diff
@@ -0,0 +1,265 @@
+   * and, if it passes, is sent to permanent storage, along with the uploaded image's blob key.
+   */
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Check to see if the user is currently logged in
+    UserService userService = backstoryUserServiceFactory.newInstance();
+    if (!userService.isUserLoggedIn()) {
+      String urlToRedirectToAfterUserLogsIn = "/analyze-image";
+      String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+      response.sendRedirect(loginUrl);
+
+    } else {
+      // Get user identification to store alongside their backstory and image
+      String userEmail = userService.getCurrentUser().getEmail();
+
+      // The blobKeyString of the image will be used to serve the image back to the front-end.
+      BlobstoreManager blobstoreManager = blobstoreManagerFactory.newInstance();
+      final String blobKeyString =
+          blobstoreManager.getUploadedFileBlobKeyString(request, "image-upload");
+      // The raw byte array representing the image will be used for image analytics.
+      final byte[] bytes = blobstoreManager.getBlobBytes(request, "image-upload");
+
+      // Validate that an image was actually uploaded.
+      if (bytes == null || blobKeyString == null) {
+        // Redirect back to the HTML page.
+        response.sendError(400, "Please upload a valid image.");
+
+      } else {
+        // Generate a list of AnnotatedImages, with each annotatedImage consisting of an image with
+        // labels.
+        ImagesManager imagesManager = imagesManagerFactory.newInstance();
+        List<byte[]> imagesAsByteArrays = new ArrayList<>();
+        imagesAsByteArrays.add(bytes);
+        List<AnnotatedImage> annotatedImages =
+            imagesManager.createAnnotatedImagesFromImagesAsByteArrays(imagesAsByteArrays);
+        // Currently, Backstory only supports single image uploads.
+        // which is why we only get the first annotatedImage element here from annotatedImages.
+        AnnotatedImage annotatedImage = annotatedImages.get(0);
+        List<String> descriptions = annotatedImage.getLabelDescriptions();
+
+        PromptManager promptManager = new PromptManager(descriptions);
+        // The delimiter for the MVP prompt will be tentatively be " and ", ex:
+        // "<label1> and <label2> and <label3>"
+        String prompt = promptManager.generatePrompt(" and ");
+
+        // Tentative backstory generation parameters: a 200 word-long story, with a .7 temperature.
+        StoryManager storyManager = storyManagerFactory.newInstance(prompt, 200, .7);
+
+        // We loop until a story succesfully generates, this is necessary because of an unavoidable
+        // memory leak in the GPT2 container which causes generation to occasionaly fail.
+        String rawBackstory = "";
+        Boolean generateTextFailed = true;
```

Comment:
Could this extra variable be avoided by either calling `break` within the try-block, or checking the content of `rawBackstory`?

Reply:
Sure, I'll update the logic to follow this.